airbyte-cdk/python/unit_tests/sources/test_source.py

```diff
@@ -462,3 +473,116 @@ def test_source_config_transform_and_no_transform(abstract_source, catalog):
     assert len(records) == 2
     assert [r.record.data for r in records] == [{"value": "23"}, {"value": 23}]
+
+
+def test_read_default_http_availability_strategy_stream_available(catalog, mocker):
+    mocker.patch.multiple(HttpStream, __abstractmethods__=set())
+    mocker.patch.multiple(Stream, __abstractmethods__=set())
+
+    class MockHttpStream(MagicMock, HttpStream):
+        url_base = "http://example.com"
+        path = "/dummy/path"
+        get_json_schema = MagicMock()
+
+        def supports_incremental(self):
+            return True
+
+        def __init__(self, *args, **kvargs):
+            MagicMock.__init__(self)
+            HttpStream.__init__(self, *args, kvargs)
+            self.read_records = MagicMock()
+
+    class MockStream(MagicMock, Stream):
+        page_size = None
+        get_json_schema = MagicMock()
+
+        def __init__(self, *args, **kvargs):
+            MagicMock.__init__(self)
+            self.read_records = MagicMock()
+
+    streams = [MockHttpStream(), MockStream()]
+    http_stream, non_http_stream = streams
+    assert isinstance(http_stream, HttpStream)
+    assert not isinstance(non_http_stream, HttpStream)
+
+    assert isinstance(http_stream.availability_strategy, HttpAvailabilityStrategy)
+    assert non_http_stream.availability_strategy is None
+
+    # Add an extra record for the default HttpAvailabilityStrategy to pull from
+    # during the try: next(records) check, since we are mocking the return value
+    # and not re-creating the generator like we would during actual reading
+    http_stream.read_records.return_value = iter([{"value": "test"}] + [{}] * 3)
+    non_http_stream.read_records.return_value = iter([{}] * 3)
+
+    source = MockAbstractSource(streams=streams)
+    logger = logging.getLogger(f"airbyte.{getattr(abstract_source, 'name', '')}")
+    records = [r for r in source.read(logger=logger, config={}, catalog=catalog, state={})]
+    # 3 for http stream and 3 for non http stream
+    assert len(records) == 3 + 3
+    assert http_stream.read_records.called
+    assert non_http_stream.read_records.called
+
+
+def test_read_default_http_availability_strategy_stream_unavailable(catalog, mocker, caplog):
+    mocker.patch.multiple(Stream, __abstractmethods__=set())
+
+    class MockHttpStream(HttpStream):
+        url_base = "https://test_base_url.com"
+        primary_key = ""
+
+        def __init__(self, **kwargs):
+            super().__init__(**kwargs)
+            self.resp_counter = 1
+
+        def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+            return None
+
+        def path(self, **kwargs) -> str:
+            return ""
+
+        def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+            stub_response = {"data": self.resp_counter}
+            self.resp_counter += 1
+            yield stub_response
+
+    class MockStream(MagicMock, Stream):
```

Comment:
nit: `MagicMock` is also available under `mocker.MagicMock` (allows lighter imports)